File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/core/types/AngleValue.swift
//
//  AngleValue.swift
//  StyleConverterTest — StyleEngine/core/types
//
//  CSS angles. Fixture: examples/primitives/angles.json.
//  IR shape: { "deg": <normalized>, "original"?: { "v": <n>, "u": "DEG|RAD|GRAD|TURN" } }
//  `original` is omitted when the input was already in degrees.
//
//  We normalise everything to degrees so downstream renderers work in a
//  single unit. The original unit is intentionally discarded — iOS does not
//  need it for rendering.
//

import Foundation

// Single-field struct keeps the API surface tiny and Equatable-friendly.
struct AngleValue: Equatable {
    let degrees: Double
}

// Never throws, never returns nil. Quirk #7 (angle-as-list) handled by
// callers that expect arrays — this extractor operates on a single IRValue.
func extractAngle(_ value: IRValue?) -> AngleValue? {
    guard let value = value else { return nil }

    // Object form — the canonical shape.
    if case .object(let o) = value {
        if let deg = o["deg"]?.doubleValue {
            return AngleValue(degrees: deg)
        }
        // Legacy / defensive: some generators emitted `degrees` instead of `deg`.
        if let deg = o["degrees"]?.doubleValue {
            return AngleValue(degrees: deg)
        }
        // Only `original` is populated — resolve the unit ourselves.
        if let original = o["original"]?.objectValue,
           let v = original["v"]?.doubleValue,
           let u = original["u"]?.stringValue {
            return AngleValue(degrees: convertToDegrees(v, unit: u))
        }
        return nil
    }

    // Bare number — treat as degrees (matches converter's DEG omission rule).
    if let d = value.doubleValue { return AngleValue(degrees: d) }
    return nil
}

// Convert a raw (value, unit) pair to degrees. Unit tags match the Kotlin
// LengthUnit-for-angles enum (DEG/RAD/GRAD/TURN).
private func convertToDegrees(_ v: Double, unit: String) -> Double {
    switch unit.uppercased() {
    case "DEG":  return v
    case "RAD":  return v * 180.0 / .pi
    case "GRAD": return v * 0.9          // 400 grad == 360 deg.
    case "TURN": return v * 360.0
    default:     return v                // Unknown — assume already degrees.
    }
}

File: testing/web/src/style/engine/core/types/AngleValue.ts
// AngleValue.ts - Angle primitive extractor.
// Handles the shape documented in examples/primitives/angles.json:
//   { deg: <normalized>, original?: { v:N, u:'DEG'|'RAD'|'GRAD'|'TURN' } }
// 'original' is omitted when the source unit was already 'deg'.
// Quirk #7: Some properties (transitions, filter drop-shadow) can carry lists of angles,
// but per-element extraction is still the single-object form below.

// Normalised angle value — always in degrees so downstream code can rotate() directly.
export interface AngleValue { degrees: number }

// Convert a raw unit+magnitude pair into degrees. Returns null on unrecognised units.
function unitToDegrees(v: number, u: unknown): number | null {
  if (typeof u !== 'string') return null;                           // guard
  switch (u.toUpperCase()) {                                        // IR uppercases but accept any case
    case 'DEG':  return v;                                          // already canonical
    case 'RAD':  return v * (180 / Math.PI);                        // radians -> degrees
    case 'GRAD': return v * 0.9;                                    // 400grad == 360deg
    case 'TURN': return v * 360;                                    // one turn == 360deg
    default:     return null;                                       // unknown unit: caller treats as failure
  }
}

// Main entrypoint. Returns null on parse failure (caller decides default).
export function extractAngle(data: unknown): AngleValue | null {
  if (data === null || data === undefined) return null;

  // Bare number: assume degrees (compose/svg conventions).
  if (typeof data === 'number') return { degrees: data };

  if (typeof data !== 'object') return null;                        // strings/other types not accepted here
  const obj = data as Record<string, unknown>;

  // Canonical shape: { deg: N } already pre-normalised by the Kotlin parser.
  if (typeof obj.deg === 'number') return { degrees: obj.deg };

  // Fall-back: read the 'original' wrapper and convert from the original unit.
  if (obj.original && typeof obj.original === 'object') {
    const orig = obj.original as Record<string, unknown>;
    if (typeof orig.v === 'number') {
      const deg = unitToDegrees(orig.v, orig.u);
      if (deg !== null) return { degrees: deg };
    }
  }

  // Defensive: raw {v,u} without 'original' wrapper.
  if (typeof obj.v === 'number') {
    const deg = unitToDegrees(obj.v, obj.u);
    if (deg !== null) return { degrees: deg };
  }

  return null;                                                      // nothing recognised
}
